redash/permissions.py

```diff
@@ -16,13 +16,17 @@
 
 def has_access(obj, user, need_view_only):
     if hasattr(obj, 'api_key') and user.is_api_user():
-        return has_access_to_object(obj, user, need_view_only)
+        return has_access_to_object(obj, user.id, need_view_only)
     else:
         return has_access_to_groups(obj, user, need_view_only)
 
 
-def has_access_to_object(obj, user, need_view_only):
-    return (obj.api_key == user.id) and need_view_only
+def has_access_to_object(obj, api_key, need_view_only):
+    if obj.api_key == api_key:
+        return need_view_only
+    else:
+        # check if api_key belongs to a dashboard containing this query
+        return api_key in obj.dashboard_api_keys and need_view_only
```

Comment:
This needs tests.

Reply:
4f742fb

Reply:
We might use `has_access_to_object` for other objects, in which case calling `dashboard_api_keys` will fail. Maybe add some more ifs?

Reply:
Not sure if I see this as premature or just general defensive programming, but anyway, 4098cf2.